### DISCUSSION TITLE
Access backups using Python

### DISCUSSION THREAD

**Discussion by nlindblo** [longhorn/longhorn#5686] [General]
Hi,

I am trying to list backups of a volume using python and have found the snippet below in the test scripts which is supposed to do this. I am accessing the API from outside of Kubernetes via a Traefik load balancer but when my script executes

backupList().data

I get a timeout error: 

requests.exceptions.ConnectionError: HTTPConnectionPool(host='10.0.4.29', port=9500): Max retries exceeded with url: /v1/backupvolumes/pvc-72013915-c662-456e-80c2-4aa0240ca171?action=backupList (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x1033c6c40>: Failed to establish a new connection: [Errno 60] Operation timed out'))

10.0.4.29 is a kubernetes node IP, does the backupList function need to be executed from a pod where the node is accessible by node IP or is there any way around this ? If anyone have a script or know how I can list and delete a backup using python from outside the cluster would be appreciated 

def find_backup(client, vol_name, snap_name):

    def find_backup_volume():
        bvs = client.list_backupVolume()
        for bv in bvs:
            if bv.name == vol_name:
                return bv
        return None

    bv = None
    for i in range(120):
        if bv is None:
            bv = find_backup_volume()
        if bv is not None:
            backups = bv.backupList().data
            for b in backups:
                if b.snapshotName == snap_name:
                    return bv, b
        time.sleep(RETRY_BACKUP_INTERVAL)
    assert False, "failed to find backup for snapshot " + snap_name + \
                  " for volume " + vol_name

**Comment by mantissahz** [comment]
Hi @nlindblo,
I just test the codes you provided and it seems to work well (`bv.backupList().data` is ok)
And `client.list_backupVolume()` would be executed first and passed.
Then `bv.backupList().data` was failed.
That's a bit weird.
Is there any abnormal log in the `longhron-manager` pod of this `10.0.4.29` IP?

**Comment by mantissahz** [comment]
Could you see the backup volume `pvc-72013915-c662-456e-80c2-4aa0240ca171` on the Page `Backup` of the UI by the browser?

**Comment by nlindblo** [comment]
Just to clarify my environment here, I am running the code from my laptop on subnet 192.168.100.0/24, I have a Traefik Ingress on this same subnet that the Longhorn UI has an IngressRoute to. The 10.0.4.29 IP address is the IP address of one of the longhorn-manager pods inside the cluster that I have no route to from my laptop and hence the connection timeout. 

Having looked at the documentation again, it states that I need to setup a port forward to my local machine

`kubectl port-forward services/longhorn-frontend 8080:http -n longhorn-system`

So possibly its by the design that connecting the endpoint via an Ingress is not supposed to work.

This works
```
def list_volumes():
    client = longhorn.Client(url='http://longhorn.mydomain.cloud/v1')
    volumes = client.list_volume()
    for volume in volumes:
        print(volume.name)
```


and gives me a list of volumes as expected
```
pvc-651db684-dc06-42cc-8f2f-34fe752d4f8b
pvc-72013915-c662-456e-80c2-4aa0240ca171
pvc-77b086d2-8082-4dd0-ab74-20bd2ef2fc10
pvc-81d25a70-7933-4610-821e-fc796704c213
pvc-81f7f935-857c-4648-ae6a-b4ed93f6aa99
pvc-c679c293-2b37-4f7c-8b81-361c7d2599a7
pvc-cefc3caa-830e-4639-b0ec-778ea7899281
pvc-e093e706-87e9-4627-9d28-92fd64106677
pvc-e42d48a6-24f8-4bc8-9079-7aedebb8bd82
pvc-e992a048-fd67-4bfb-8afb-f64264f27508
pvc-ebf61157-31d3-4af4-bab5-bdc89f0cbc25
pvc-ee4006a3-4fe3-44d1-adbe-51d58e56ae8f
```

This doesn't work
```
  def list_backups():
    volume_name = "pvc-651db684-dc06-42cc-8f2f-34fe752d4f8b"
    client = longhorn.Client(url='http://longhorn.mydomain.cloud/v1')
    backup_volume = client.by_id_backupVolume(volume_name)

    backups = backup_volume.backupList().data
    for backup in backups:
        print(backup)
```
And gives me the connection error to the longhorn-manager pod.

If anyone can confirm my assumptions about how this is supposed to work (as the documentation of the API and Python client is a bit light) or any thoughts of a workaround would be appreciated.

**Reply by mantissahz** [reply]
Is the volume `pvc-651db684-dc06-42cc-8f2f-34fe752d4f8b` in `client.list_backupVolume()` list?
Did you do backups for this volume `pvc-651db684-dc06-42cc-8f2f-34fe752d4f8b`?
Could you show the result of `client.list_backupVolume()`?

And could you describe your environment more? OS, kubernetes cluster, Longhorn version?

**Comment by nlindblo** [comment]
Yes, the volume is in the backup volume list

```
def list_backup_volumes():
  client = longhorn.Client(url='http://longhorn.mydomain.cloud/v1')
  backupvolumes = client.list_backupVolume()
  for backupvolume in backupvolumes:
    print(backupvolume.name)

pvc-651db684-dc06-42cc-8f2f-34fe752d4f8b
pvc-72013915-c662-456e-80c2-4aa0240ca171
pvc-77b086d2-8082-4dd0-ab74-20bd2ef2fc10
pvc-81d25a70-7933-4610-821e-fc796704c213
pvc-81f7f935-857c-4648-ae6a-b4ed93f6aa99
pvc-c679c293-2b37-4f7c-8b81-361c7d2599a7
pvc-cefc3caa-830e-4639-b0ec-778ea7899281
pvc-e093e706-87e9-4627-9d28-92fd64106677
pvc-e42d48a6-24f8-4bc8-9079-7aedebb8bd82
pvc-e992a048-fd67-4bfb-8afb-f64264f27508
pvc-ebf61157-31d3-4af4-bab5-bdc89f0cbc25
pvc-ee4006a3-4fe3-44d1-adbe-51d58e56ae8f

```

And yes, the volume has backups, which I can actually list in a different way with this code

```
def listbackups_for_volume():
  client = longhorn.Client(url='http://longhorn.mydomain.cloud/v1')
  volume = client.by_id_volume("pvc-651db684-dc06-42cc-8f2f-34fe752d4f8b")

  for mybup in get_backups(client, volume):
    print("ID " + mybup.id + " State " + mybup.state)
    print("===")

def get_backups(client, volume):
  client = longhorn.Client(url='http://longhorn.mydomain.cloud/v1')
  backups = []

  bups = volume.backupStatus
  for bup in bups:
    backups.append(bup)

  return(backups)
```
Gives
```
ID backup-09f63e2df6d1426d State Completed
===
ID backup-71b3e5fe44974de0 State Completed
===
ID backup-8ff75e5593e14a45 State Completed
===
ID backup-e5715b53409640bc State Completed
===
```

My ultimate goal is to be able to delete a backup using backup_volume.backupDelete(name=backup_name) which has the same issue as backupList where the function seem to issue a python request to the pod IP on port 9500 and I thought the backupList was a simpler example

This is my backup deletion code, you can see the backup ID and backup volumes listed above (i.e. they exist)

```
def deletebackup():
  client = longhorn.Client(url='http://longhorn.mydomain.cloud/v1')
  volume = client.by_id_volume("pvc-651db684-dc06-42cc-8f2f-34fe752d4f8b")
  backup_volume = client.by_id_backupVolume(volume.name)
  backupid = "backup-71b3e5fe44974de0"

  backup_volume.backupDelete(name=backupid)
```
Gives
```
requests.exceptions.ConnectionError: HTTPConnectionPool(host='10.0.4.47', port=9500): Max retries exceeded with url: /v1/backupvolumes/pvc-651db684-dc06-42cc-8f2f-34fe752d4f8b?action=backupDelete (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x101446b50>: Failed to establish a new connection: [Errno 60] Operation timed out'))
```

I have describe my network environment already above. Kubernetes version is 1.26 running on Debian 11 and Longhorn version is 1.4.1

**Reply by mantissahz** [reply]
Could you test the python code `deleteBackup` again and create the support bundle?
The support bundle will help us analyze the root cause. Thanks.

**Reply by nlindblo** [reply]
Thanks, I have created a support bundle and it seem to collect quite a lot of data that I don't feel comfortable posting here. Are there any specific config or log files I can take out from the bundle and post here ?

**Reply by nlindblo** [reply]
Here are all the longhorn-logs, anything more needed from the bundle just let me know

[longhorn-system.zip](https://github.com/longhorn/longhorn/files/11123394/longhorn-system.zip)

The support bundle was generated directly after the timeout message following running
```
  client = longhorn.Client(url='http://longhorn.mydomain.cloud/v1')
  volume = client.by_id_volume("pvc-651db684-dc06-42cc-8f2f-34fe752d4f8b")
  backup_volume = client.by_id_backupVolume(volume.name)
  backupid = "backup-71b3e5fe44974de0"

  backup_volume.backupDelete(name=backupid)
```
error message
```
requests.exceptions.ConnectionError: HTTPConnectionPool(host='10.0.0.148', port=9500): Max retries exceeded with url: /v1/backupvolumes/pvc-651db684-dc06-42cc-8f2f-34fe752d4f8b?action=backupDelete (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x103672d30>: Failed to establish a new connection: [Errno 60] Operation timed out'))
```

**Reply by mantissahz** [reply]
Could you try to test the codes on `node1` and use the URL `http://10.0.4.189:9500/v1`?

just found the logs in the bundle. APIs should work.
```txt
longhorn-manager-t5pc9/longhorn-manager.log:1004:2023-03-31T17:04:38.517011961+02:00 10.0.0.212 - - [31/Mar/2023:15:04:38 +0000] "POST /v1/backupvolumes/pvc-651db684-dc06-42cc-8f2f-34fe752d4f8b?action=backupList HTTP/1.1" 200 4547 "http://longhorn.mystorm.cloud/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.3 Safari/605.1.15"
longhorn-manager-t5pc9/longhorn-manager.log:1005:2023-03-31T17:04:38.517269515+02:00 10.0.0.212 - - [31/Mar/2023:15:04:38 +0000] "POST /v1/backupvolumes/pvc-651db684-dc06-42cc-8f2f-34fe752d4f8b?action=backupList HTTP/1.1" 200 4547 "http://longhorn.mystorm.cloud/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.3 Safari/605.1.15"
```
